### PR TITLE
make: add error message for bad CUDA version

### LINKED ISSUE #5294: nvcc fatal : Value 'native' is not defined for option 'gpu-architecture' when compiling cuBLAS
There seems to be an issue on the Makefile that's assigning an invalid nvcc flag for gpu-architecture. I'm running on Ubuntu22.04 on a RTX4090, starting from the Docker image nvidia/cuda:12.3.1-runtime-ubuntu22.04.

Running make creates the following error:

```
root@ZEPPELIN-01:/workspace/llama.cpp# make LLAMA_CUBLAS=1
expr: syntax error: unexpected argument '070100'
expr: syntax error: unexpected argument '080100'
I llama.cpp build info: 
I UNAME_S:   Linux
I UNAME_P:   x86_64
I UNAME_M:   x86_64
I CFLAGS:    -I. -Icommon -D_XOPEN_SOURCE=600 -D_GNU_SOURCE -DNDEBUG -DGGML_USE_CUBLAS -I/usr/local/cuda/include -I/opt/cuda/include -I/targets/x86_64-linux/include -I/usr/local/cuda/targets/aarch64-linux/include  -std=c11   -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wshadow -Wstrict-prototypes -Wpointer-arith -Wmissing-prototypes -Werror=implicit-int -Werror=implicit-function-declaration -pthread -march=native -mtune=native -Wdouble-promotion 
I CXXFLAGS:  -I. -Icommon -D_XOPEN_SOURCE=600 -D_GNU_SOURCE -DNDEBUG -DGGML_USE_CUBLAS -I/usr/local/cuda/include -I/opt/cuda/include -I/targets/x86_64-linux/include -I/usr/local/cuda/targets/aarch64-linux/include  -std=c++11 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread  -march=native -mtune=native -Wno-array-bounds -Wno-format-truncation -Wextra-semi
I NVCCFLAGS: -use_fast_math --forward-unknown-to-host-compiler -arch=native -DGGML_CUDA_DMMV_X=32 -DGGML_CUDA_MMV_Y=1 -DK_QUANTS_PER_ITERATION=2 -DGGML_CUDA_PEER_MAX_BATCH_SIZE=128 
I LDFLAGS:   -lcuda -lcublas -lculibos -lcudart -lcublasLt -lpthread -ldl -lrt -L/usr/local/cuda/lib64 -L/opt/cuda/lib64 -L/targets/x86_64-linux/lib -L/usr/local/cuda/targets/aarch64-linux/lib -L/usr/lib/wsl/lib 
I CC:        cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
I CXX:       g++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0

nvcc -I. -Icommon -D_XOPEN_SOURCE=600 -D_GNU_SOURCE -DNDEBUG -DGGML_USE_CUBLAS -I/usr/local/cuda/include -I/opt/cuda/include -I/targets/x86_64-linux/include -I/usr/local/cuda/targets/aarch64-linux/include  -std=c++11 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread  -use_fast_math --forward-unknown-to-host-compiler -arch=native -DGGML_CUDA_DMMV_X=32 -DGGML_CUDA_MMV_Y=1 -DK_QUANTS_PER_ITERATION=2 -DGGML_CUDA_PEER_MAX_BATCH_SIZE=128  -Wno-pedantic -Xcompiler "-Wno-array-bounds" -c ggml-cuda.cu -o ggml-cuda.o
nvcc fatal   : Value 'native' is not defined for option 'gpu-architecture'
make: *** [Makefile:429: ggml-cuda.o] Error 1
```

The same issue seems to happen in whisper.cpp, by the way -- I also created an issue report there.

The issue seems to be resolved by simply changing the flag on the Makefile from 'native' to 'all'. However, I still cannot make llama.cpp to compile with the GPU for some reason.

### PR DESCRIPTION
**Pull request by JohannesGaessler**
Fixes https://github.com/ggerganov/llama.cpp/issues/5294 in the sense that there is an error message to tell the user what the problem is. The root problem is that for old CUDA versions `-arch=native` is not supported. However, compiling for multiple compute capabilities like with cmake also has drawbacks: the compilation time would increase and nvcc does not seem to natively support compiling for multiple specific CUDA architectures in parallel. So you would instead have to compile multiple .cubin files and then link them together. And since this is only an issue for old CUDA versions I think it's fine to just add an error message that tells users how to fix it.

Currently the error message tells the user to set `CUDA_DOCKER ARCH`. The people with the problem were indeed using Docker but in principle you could have this exact same problem with bare metal. Should we rename the argument to something like `CUDA_ARCH`?

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -569,6 +569,14 @@ $(info I CC:        $(shell $(CC)   --version | head -n 1))
 $(info I CXX:       $(shell $(CXX)  --version | head -n 1))
 ifdef LLAMA_CUBLAS
 $(info I NVCC:      $(shell $(NVCC) --version | tail -n 1))
+CUDA_VERSION := $(shell nvcc --version | grep -oP 'release (\K[0-9]+\.[0-9])')
+ifeq ($(shell echo "$(CUDA_VERSION) < 11.7" | bc),1)
```
Comment (cebtenzzre) [inline Makefile:573]: awk is perhaps a more standard compile-time dependency than bc:
```suggestion
ifeq ($(shell awk -v "v=$(CUDA_VERSION)" 'BEGIN { print (v < 11.7) }'),1)
```
e.g. Arch Linux has awk in base and base-devel, while bc is in neither.

**Comment by cebtenzzre** [comment]
> Should we rename the argument to something like `CUDA_ARCH`?

I am a proponent of this change, calling it CUDA_DOCKER_ARCH never made sense to me.

File: Makefile (hunk 1 above)
Comment (JohannesGaessler) [reply Makefile:573]: Why not like this?

```suggestion
ifeq ($(shell awk 'BEGIN { print ($(CUDA_VERSION) < 11.7) }'),1)
```

To me this is both more intuitive and simpler.

Comment (cebtenzzre) [reply Makefile:573]: That works too, I just like to be in the habit of keeping data and code separate, if only so you don't get a syntax error (or worse) if the value contains unexpected characters that don't get escaped properly.

Comment (JohannesGaessler) [reply Makefile]: Okay, the point with syntax errors makes sense.

**Comment by ggerganov** [comment]
> Should we rename the argument to something like CUDA_ARCH?

Btw, on similar note, we should rename `LLAMA_CUBLAS` to `LLAMA_CUDA` and `GGML_USE_CUBLAS` to `GGML_USE_CUDA` at some point

**Review by slaren — APPROVED** [review]

**Pull request merged**